staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go

```diff
@@ -23,35 +23,75 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	utiljson "k8s.io/apimachinery/pkg/util/json"
+	kjson "sigs.k8s.io/json"
 )
 
 // GetObjectMeta does conversion of JSON to ObjectMeta. It first tries json.Unmarshal into a metav1.ObjectMeta
 // type. If that does not work and dropMalformedFields is true, it does field-by-field best-effort conversion
 // throwing away fields which lead to errors.
 func GetObjectMeta(obj map[string]interface{}, dropMalformedFields bool) (*metav1.ObjectMeta, bool, error) {
+	meta, found, _, err := GetObjectMetaWithOptions(obj, ObjectMetaOptions{
+		DropMalformedFields: dropMalformedFields,
+	})
+	return meta, found, err
+}
+
+// objectMeta is a struct used for umarshaling wrapped metadata bytes.
+type objectMeta struct {
+	Metadata *metav1.ObjectMeta `json:"metadata"`
+}
+
+// ObjectMeta proved the options for how GetObjectMeta should retrieve the object meta.
+type ObjectMetaOptions struct {
+	DropMalformedFields bool
+	ReturnPruned        bool
+	ParentPath          string
+}
+
+// GetObjectMetaWithOptions allows for strict unmarshaling of the metadata bytes, returning any unknown field paths
+// it encounters
+func GetObjectMetaWithOptions(obj map[string]interface{}, opts ObjectMetaOptions) (*metav1.ObjectMeta, bool, []error, error) {
 	metadata, found := obj["metadata"]
 	if !found {
-		return nil, false, nil
+		return nil, false, nil, nil
 	}
 
 	// round-trip through JSON first, hoping that unmarshalling just works
-	objectMeta := &metav1.ObjectMeta{}
-	metadataBytes, err := utiljson.Marshal(metadata)
+	objectMeta := &objectMeta{}
+	wrappedMetadataBytes, err := utiljson.Marshal(map[string]interface{}{"metadata": metadata})
 	if err != nil {
-		return nil, false, err
+		return nil, false, nil, err
 	}
-	if err = utiljson.Unmarshal(metadataBytes, objectMeta); err == nil {
-		// if successful, return
-		return objectMeta, true, nil
+	if opts.ReturnPruned {
+		var strictErrs []error
+		strictErrs, err = kjson.UnmarshalStrict(wrappedMetadataBytes, objectMeta)
+		if err == nil {
```

Comment:
hoisting from the outdated comment thread at https://github.com/kubernetes/kubernetes/pull/109494#discussion_r912262730:

By only paying attention to strict errors when err is nil here, I don't think this is doing the right thing in combination with dropMalformedFields... updating one of the unit tests that looks at dropMalformedFields to expect it to also correctly track unknown field paths fails:

```diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/algorithm_test.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/algorithm_test.go
index f8eb2ebae43..2630cbef4fb 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/algorithm_test.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/algorithm_test.go
@@ -412,7 +412,8 @@ func TestCoerce(t *testing.T) {
     "kind": 42,
     "metadata": {
       "name": "instance",
+      "unspecified": "bar",
	  "namespace": ["abc"],
       "labels": {
         "foo": 42
       }
@@ -447,7 +448,7 @@ func TestCoerce(t *testing.T) {
     }
   }
 }
-`},
+`, expectedUnknownFields: []string{"pruned.metadata.unspecified"}},
 		{name: "invalid metadata type, with dropInvalidFields=true", json: `
 {
   "apiVersion": "foo/v1",
```

That said, I'm not 100% sure we *have* to handle both ReturnPruned=true and DropMalformedFields=true at the same time... we only care about knowing about pruned fields when decoding an incoming request... can you verify that we aren't setting DropMalformedFields=true in that case? I think we may only be setting DropMalformedFields=true when decoding from etcd. If that's the case, we could maybe make DropMalformedFields and ReturnPruned mutually exclusive

Reply:
oh... I think supporting both may be as simple as teaching the innermost Unmarshal about strict validation:

```diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go
index 35e6e827687..b9e2c1f47df 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go
@@ -99,13 +99,21 @@ func GetObjectMetaWithOptions(obj map[string]interface{}, opts ObjectMetaOptions
 	// each iteration preserving the old key-values.
 	accumulatedObjectMeta := &metav1.ObjectMeta{}
 	testObjectMeta := &metav1.ObjectMeta{}
+	var unknownFields []string
 	for k, v := range metadataMap {
 		// serialize a single field
 		if singleFieldBytes, err := utiljson.Marshal(map[string]interface{}{k: v}); err == nil {
 			// do a test unmarshal
 			if utiljson.Unmarshal(singleFieldBytes, testObjectMeta) == nil {
 				// if that succeeds, unmarshal for real
-				utiljson.Unmarshal(singleFieldBytes, accumulatedObjectMeta)
+				if opts.ReturnPruned {
+					strictErrs, err = kjson.UnmarshalStrict(wrappedMetadataBytes, objectMeta)
+					if len(strictErrs) > 0 {
+						... compute unknown fields
+					}
+				} else {
+					utiljson.Unmarshal(singleFieldBytes, accumulatedObjectMeta)
+				}
 			}
 		}
 	}
```

if it's simple to support both simultaneously, let's just do that... it might be easier than reasoning about whether we ever use both options simultaneously